### ISSUE TITLE
Vocabulary class for preprocess

### ISSUE THREAD

**Issue by faneshion** [faneshion/MatchZoo#206]
Here, we need to design a `Vocabulary` class in the preprocess. The `Vocabulary` class is to map a list of words into a list of word-ids. Maybe there can be a 'fit' function to store '_word_dict', and a 'transform' function to map words to ids. We can discuss it more~ @uduse @pl8787 @yangliuy @bwanglzu

**Comment by bwanglzu** [comment]
Elaborate a little bit more on `word_dict`?

**Comment by uduse** [comment]
`StatefulProcessorUnit` suffices. A quick draft here: https://github.com/faneshion/MatchZoo/blob/feature/vocabulary/matchzoo/preprocessors.py#L171

**Comment by faneshion** [comment]
The __missing__ function is a good idea to deal with OOV words. Can the `fit` function increasingly build the `token_index` and `index_token`? In most times, we would call the `fit` several times to build the final `vocabulary`.

**Comment by uduse** [comment]
@faneshion Yes that's a common usage. However, a `fit` usually resets the processor's inner state. 

e.g. sklearn's `LabelEncoder`.
![image](https://user-images.githubusercontent.com/4717005/42071225-de14aa02-7b8c-11e8-910d-c189a2fd078b.png)

It's important to keep the interface consistent. Either we don't reset inner state for all stateful processors, or we implements such incremental fitting for all of them. The latter one is harder since one needs to keep track of more things. However, on the user's end, `fit` multiple times isn't much more convenient that collect all tokens into a list and do a single `fit` (only takes two more lines).

Another solution would be not using `StatefulProcessorUnit` but a dedicated class.

**Comment by faneshion** [comment]
We can leave the incremental preprocess to the future version. Right now, it can be satisfied for the current usage. You can pull request this code to be checked further more.
